### PR TITLE
Initial checkin of potential fix

### PR DESCRIPTION
**Pull request by cahnk**

### CODE REVIEW

**Comment by cahnk** [comment]
This update correctly detects terraform files in the root and modules directories, and updates the README files at both locations with tfdocs and newlines at the end.  It does have some potentially unwanted but harmless side-effects.  

1.  If the root directory does not have a README file, then it creates one with a tfdocs section.
2. If the root directory has a README file and no terraform files, then it still adds a tfdocs section to the README file.
3. If a modules subfolder contains terraform files and no README file, then it creates a README file in the subfolder with a tfdocs section.  This is probably the desired behavior most, if not all of the time.  

These side-effects are due to the nature in which terraform-docs handles recursive updates.  It always assumes there are terraform files in the root and modules directories, and that all of these locations should have README files with tfdocs sections.

File: Makefile [hunk 1]
```diff
@@ -388,9 +388,9 @@ docs/%: export TFDOCS_RECURSIVE ?= $(if $(wildcard $(TFDOCS_PATH)/modules),--rec
 docs/%: export TFDOCS_TEMPLATE ?= <!-- BEGIN TFDOCS -->\n{{ .Content }}\n\n<!-- END TFDOCS -->
 docs/%: export TFDOCS_FILE ?= README.md
 docs/%: export TFDOCS_PATH ?= .
-docs/%: export README_FILES ?=  $(if $(wildcard $(TFDOCS_FILE)),true,)
-docs/%: export TF_FILES ?= $(if $(wildcard $(TFDOCS_PATH)/*.tf),true,)
-docs/%: export HCL_FILES ?= $(if $(wildcard $(TFDOCS_PATH)/*.pkr.hcl),true,)
+docs/%: export README_FILES ?=  $(if $(wildcard $(TFDOCS_FILE)),true,$(if $(wildcard modules/*/$(TFDOCS_FILE)),true,))
+docs/%: export TF_FILES ?= $(if $(wildcard $(TFDOCS_PATH)/*.tf),true,$(if $(wildcard modules/*/*.tf),true,))
+docs/%: export HCL_FILES ?= $(if $(wildcard $(TFDOCS_PATH)/*.pkr.hcl),true,$(if $(wildcard /modules/*/*.pkr.hcl),true,))
```
Comment (lorengordon) [inline Makefile:393]: The structure for packer projects is a bit different. I don't think we need to make this change at this time...

File: Makefile [hunk 2]
```diff
@@ -403,6 +403,10 @@ docs/generate: | terraform/format
 		echo "Adding newline to the end of $(TFDOCS_FILE) file"; \
 		echo "" >> $(TFDOCS_FILE); \
 	fi
+	@if [ -n $(README_FILES) ] && [ "$$(tail -c1 modules/*/$(TFDOCS_FILE))" != "$('\n')" ]; then \
```
Comment (lorengordon) [inline Makefile:406]: Can you discuss how this works a bit? Biggest concern: when there are multiple modules, and one has a newline and one does not. Does this add a newline to both?

**Review by lorengordon — COMMENTED** [review]
I'm not sure I love those side effects, but we can give it a try and see how it goes...

File: Makefile (hunk 1 above)
Comment (cahnk) [reply Makefile:393]: I reverted line 393.

File: Makefile (hunk 2 above)
Comment (cahnk) [reply Makefile:406]: I just tested this with two modules subfolders and it causes an ambiguous redirect error.  I'll have to work on this some more.  Good catch.

File: Makefile [hunk 3]
```diff
@@ -388,21 +388,28 @@ docs/%: export TFDOCS_RECURSIVE ?= $(if $(wildcard $(TFDOCS_PATH)/modules),--rec
...
 docs/generate: | terraform/format
 	@[ "${TFDOCS_CMD}" ] && ( echo "[$@]: Generating docs";) || ( echo "[$@]: No docs to generate";)
 	$(TFDOCS_CMD)
-	@if [ -n $(README_FILES) ] && [ "$$(tail -c1 $(TFDOCS_FILE))" != "$('\n')" ]; then \
-		echo "Adding newline to the end of $(TFDOCS_FILE) file"; \
-		echo "" >> $(TFDOCS_FILE); \
-	fi
+	@make docs/format
+
+## Adds a newline to the end of all project markdown files if one does not already existGenerates Terraform documentation
```
Comment (lorengordon) [inline Makefile:405]: Comment has some copy pasta

```suggestion
## Adds a newline to the end of all project markdown files if one does not already exist
```

File: Makefile [hunk 4]
```diff
@@ -388,21 +388,28 @@ docs/%: export TFDOCS_RECURSIVE ?= $(if $(wildcard $(TFDOCS_PATH)/modules),--rec
...
 
 ## Generates Terraform documentation
 docs/generate: | terraform/format
 	@[ "${TFDOCS_CMD}" ] && ( echo "[$@]: Generating docs";) || ( echo "[$@]: No docs to generate";)
 	$(TFDOCS_CMD)
-	@if [ -n $(README_FILES) ] && [ "$$(tail -c1 $(TFDOCS_FILE))" != "$('\n')" ]; then \
-		echo "Adding newline to the end of $(TFDOCS_FILE) file"; \
-		echo "" >> $(TFDOCS_FILE); \
-	fi
+	@make docs/format
```
Comment (lorengordon) [inline Makefile:403]: Use the `SELF` variable for recursive make.

```suggestion
	@$(SELF) docs/format
```

File: Makefile [hunk 5]
```diff
@@ -388,21 +388,28 @@ docs/%: export TFDOCS_RECURSIVE ?= $(if $(wildcard $(TFDOCS_PATH)/modules),--rec
...
+
+## Adds a newline to the end of all project markdown files if one does not already existGenerates Terraform documentation
+docs/format:
+	@for markdown_file in $(MARKDOWN_FILES); do \
+		if [ "$$(tail -c1 $$markdown_file)" != "$('\n')" ]; then \
+			echo "Adding newline to the end of $$markdown_file file"; \
```
Comment (lorengordon) [inline Makefile:409]: The double $ should only be needed when using $(...) or ${...}. Can you confirm the output is what you expect?

```suggestion
			echo "Adding newline to the end of $${markdown_file} file"; \
```

File: Makefile [hunk 6]
```diff
@@ -388,21 +388,28 @@ docs/%: export TFDOCS_RECURSIVE ?= $(if $(wildcard $(TFDOCS_PATH)/modules),--rec
...
-	@if [ -n $(README_FILES) ] && [ "$$(tail -c1 $(TFDOCS_FILE))" != "$('\n')" ]; then \
-		echo "Adding newline to the end of $(TFDOCS_FILE) file"; \
-		echo "" >> $(TFDOCS_FILE); \
-	fi
+	@make docs/format
+
+## Adds a newline to the end of all project markdown files if one does not already existGenerates Terraform documentation
+docs/format:
+	@for markdown_file in $(MARKDOWN_FILES); do \
+		if [ "$$(tail -c1 $$markdown_file)" != "$('\n')" ]; then \
```
Comment (lorengordon) [inline Makefile:408]:
```suggestion
		if [ "$$(tail -c1 $${markdown_file})" != "$('\n')" ]; then \
```

File: Makefile (hunk 4 above)
Comment (cahnk) [reply Makefile:403]: Updated.

File: Makefile (hunk 3 above)
Comment (cahnk) [reply Makefile:405]: Updated

File: Makefile (hunk 5 above)
Comment (cahnk) [reply Makefile:409]: Here's the output of a test run with three files to update without the suggested changes:
```
README.md updated successfully
modules/test/README.md updated successfully
modules/test2/README.md updated successfully
Adding newline to the end of modules/test/README.md file
Adding newline to the end of modules/test2/README.md file
Adding newline to the end of README.md file
```
I get the same exact output after making the suggested changes.  Makefiles are still not my friend, so I'm not sure why both versions work the same.  I went ahead and left the changes in.

File: Makefile (hunk 6 above)
Comment (cahnk) [reply Makefile:408]: Updated

**Review by lorengordon — APPROVED** [review]

**Pull request merged**